pytorch_lightning/trainer/data_loading.py

```diff
@@ -133,19 +138,21 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
                 )
 
             # replace with distributed sampler
-            sampler = self._get_distributed_sampler(dataloader, shuffle)
-            dataloader = self.replace_sampler(dataloader, sampler)
+            sampler = self._get_distributed_sampler(dataloader, shuffle, mode=mode)
+            dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
 
         return dataloader
 
     @staticmethod
-    def _resolve_batch_sampler(dl_args, dataloader, sampler):
+    def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningStage] = None) -> Dict[str, Any]:
         batch_sampler = getattr(dataloader, "batch_sampler")
-        if batch_sampler is not None and type(batch_sampler) is not BatchSampler:
-            batch_sampler = type(batch_sampler)(
+        is_predicting = mode == RunningStage.PREDICTING
+        if (batch_sampler is not None and type(batch_sampler) is not BatchSampler) or mode == RunningStage.PREDICTING:
+            batch_sampler_type = IndexBatchSampler if is_predicting else type(batch_sampler)
+            batch_sampler = batch_sampler_type(
                 sampler,
                 batch_size=batch_sampler.batch_size,
-                drop_last=batch_sampler.drop_last,
+                drop_last=True if is_predicting else batch_sampler.drop_last,
```

Comment:
why this change?

Reply:
Thx, it should be `False`. We don't want to drop any batch and shouldn't trust the user.